metagraph/src/graph/representation/canonical_dbg.cpp

```diff
@@ -398,29 +406,38 @@ ::call_incoming_kmers(node_index node, const IncomingEdgeCallback &callback) con
 
     const auto &alphabet = graph_->alphabet();
 
-    if (auto fetch = parent_node_cache_.TryGet(node)) {
-        for (size_t c = 0; c < alphabet.size(); ++c) {
-            if ((*fetch)[c] != npos)
-                callback((*fetch)[c], alphabet[c]);
-        }
+    if (use_parent_child_cache_) {
+        if (auto fetch = parent_node_cache_.TryGet(node)) {
+            for (size_t c = 0; c < alphabet.size(); ++c) {
+                if ((*fetch)[c] != npos)
+                    callback((*fetch)[c], alphabet[c]);
+            }
 
-        return;
+            return;
+        }
     }
 
     SmallVector<node_index> parents(alphabet.size(), npos);
     size_t max_num_edges_left = parents.size() - has_sentinel_;
 
-    graph_->call_incoming_kmers(node, [&](node_index prev, char c) {
+    auto incoming_kmer_callback = [&](node_index prev, char c) {
         if (c != boss::BOSS::kSentinel) {
             parents[alphabet_encoder_[c]] = prev;
             --max_num_edges_left;
         }
-    });
+    };
+
+    if (const auto cache = get_extension<NodeFirstCache>()) {
+        cache->call_incoming_kmers(node, incoming_kmer_callback);
+    } else {
+        graph_->call_incoming_kmers(node, incoming_kmer_callback);
+    }
```

Comment:
I think there is an unnecessary degree of freedom: `get_extension<NodeFirstCache>` vs. `use_parent_child_cache_`. Ideally we should have a single parameter for both. See another question above.

Reply:
I was doing this for now as part of my testing. Ideally, `caches::fixed_sized_cache` should just not cache anything if the size is set to 0 (the default behaviour is to cache everything).

Reply:
I created a new PR in the caches library to support empty caches, so I can update this accordingly.

Reply:
I got rid of this variable now.